src/Scenes/GameScene/Renderers/Effects/NextProjectionMoveEffect.ts

```diff
@@ -0,0 +1,83 @@
+import * as PIXI from 'pixi.js';
+import {Effect} from './Effect';
+import {FastPool, Pool} from 'evidently-data-structures';
+import {Protagonist} from '../../../../GameLogic/Entities/Protagonist';
+import Constants from '../../../../Core/Constants';
+import {PlayerActionUtils} from '../../../../GameLogic/Enums';
+import {Direction8Utils} from '../../../../GameLogic/Enums/Direction8';
+import {GfxConstants} from '../../../../Core/Constants/GfxConstants';
+import {TextureStore} from 'evidently-pixi';
+
+const initialAlpha = 0.7;
+const finalAlpha = 0.3;
+const slideLength = 1000;
+const effectBreak = 500;
+const cycleLength = slideLength + effectBreak;
+
+export class NextProjectionMoveEffect extends PIXI.Sprite implements Effect {
+	private static readonly _pool: Pool<NextProjectionMoveEffect>;
+
+	public static getOne(projection: Protagonist, textureStore: TextureStore): NextProjectionMoveEffect {
+		const effect = pool.getOne();
+
+		effect.init(projection, textureStore);
+
+		return effect;
+	}
+
+	private _projection!: Protagonist;
+
+	private _fromX!: number;
+
+	private _fromY!: number;
+
+	private _toX!: number;
+
+	private _toY!: number;
+
+	private _timer = 0;
+
+	public update(timePassed: number): void {
+		this._timer = (this._timer + timePassed) % cycleLength;
+
+		const slideFraction = 1 - (1 - Math.min(this._timer / slideLength, 1)) ** 2;
+		this.x = this._fromX + (this._toX - this._fromX) * slideFraction | 0;
+		this.y = this._fromY + (this._toY - this._fromY) * slideFraction | 0;
+		this.alpha = initialAlpha + (finalAlpha - initialAlpha) * slideFraction;
+	}
+
+	public release(): void {
+		this._projection = undefined!;
```

Comment:
What does it mean if undefined is followed by an exclamation mark?

Reply:
It's a [type assertion operator](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-type-assertions). It's used to tell the compiler to disregard the strict null/undefined checking.

The broader context: because this class is instantiated with a pool, the real type of `_projection` is always going to be `Protagonist | undefined`. But I don't want to have to bother with checking if `_projection` is undefined or not, because the contract of this class is "if this class is used it HAS that property set to a valid value". That's not something that can be guaranteed with code so I have to override compiler's checks in two places to make it so. 
Thus, you may notice the definition of that field also has the exclamation mark - it's telling the compiler "yea well that variable is undefined but don't worry, I've got this".
And this specific case is me telling the compiler "I know I am setting a non-nullable variable to undefined, that's okay, I know what I am doing".

That being said, I think this field can be completely dropped anyway. I'll change it.

---

As a side note, I do the same in the definition of the `fromX` etc properties. They are supposed to be `number`, but because I don't set them in the constructor they start as `undefined`. But I know that before the instance is going to be used, it'll pass through the call to `initialization()` that'll ensure that either they're set or that's meaningless.

And now that I think about it I don't like it, I'll add a constructor that sets them to some value.